modules/mobile-agents/android/client/src/org/wso2/mdm/agent/services/Operation.java

```diff
@@ -0,0 +1,899 @@
+				ringDevice(operationCode);
+				break;
+			case Constants.OPERATION_MUTE:
+				muteDevice(operationCode);
+				break;
+			case Constants.OPERATION_WEBCLIP:
+				createWebClip(operationCode, operationData);
+				break;
+			case Constants.OPERATION_PASSWORD_POLICY:
+				setPasswordPolicy(operationCode, operationData);
+				break;
+			case Constants.OPERATION_INSTALL_GOOGLE_APP:
+				installGooglePlayApp(operationCode, operationData);
+				break;
+			case Constants.OPERATION_CHANGE_LOCK_CODE:
+				changeLockCode(operationCode, operationData);
+				break;
+			case Constants.OPERATION_POLICY_BUNDLE:
+				setPolicyBundle(operationCode);
+				break;
+			case Constants.OPERATION_ENTERPRISE_WIPE_DATA:
+				enterpriseWipe(operationCode);
+				break;
+			case Constants.OPERATION_BLACKLIST_APPS:
+				blacklistApps(operationCode, operationData);
+				break;
+			default:
+				Log.e(TAG, "Invalid operation code receieved");
+				break;
+		}
+
+		return resultBuilder.getResultPayload();
+	}
+
+	/**
+	 * Retrieve device information.
+	 * @param code - Operation code.
+	 */
+	public void getDeviceInfo(String code) {
+		DeviceInfo deviceInfo = new DeviceInfo(context.getApplicationContext());
+		DeviceState phoneState = DeviceStateFactory.getDeviceState(context.getApplicationContext(),
+	                                                           				deviceInfo.getSdkVersion());
+		GPSTracker gps = new GPSTracker(context.getApplicationContext());
+		JSONObject result = new JSONObject();
+		JSONObject batteryInfo = new JSONObject();
+		JSONObject internalMemoryInfo = new JSONObject();
+		JSONObject externalMemoryInfo = new JSONObject();
+		JSONObject locationInfo = new JSONObject();
+		double latitude;
+		double longitude;
+		
+		try {
+			latitude = gps.getLatitude();
+			longitude = gps.getLongitude();
+			int batteryLevel = (int) Math.floor(phoneState.getBatteryLevel());
+			batteryInfo.put(BATTERY_INFO_TAG_LEVEL, batteryLevel);
+
+			internalMemoryInfo.put(MEMORY_INFO_TAG_TOTAL, phoneState.getTotalInternalMemorySize());
+			internalMemoryInfo.put(MEMORY_INFO_TAG_AVAILABLE,
+			                       phoneState.getAvailableInternalMemorySize());
+			externalMemoryInfo.put(MEMORY_INFO_TAG_TOTAL, phoneState.getTotalExternalMemorySize());
+			externalMemoryInfo.put(MEMORY_INFO_TAG_AVAILABLE,
+			                       phoneState.getAvailableExternalMemorySize());
+			locationInfo.put(LOCATION_INFO_TAG_LATITUDE, latitude);
+			locationInfo.put(LOCATION_INFO_TAG_LONGITUDE, longitude);
+
+			result.put(BATTERY_INFO_TAG, batteryInfo);
+			result.put(MEMORY_INFO_TAG_INTERNAL, internalMemoryInfo);
+			result.put(MEMORY_INFO_TAG_EXTERNAL, externalMemoryInfo);
+			if (latitude != 0 && longitude != 0) {
+				result.put(LOCATION_INFO_TAG, locationInfo);
+			}
+			result.put(NETWORK_OPERATOR_TAG, deviceInfo.getNetworkOperatorName());
+
+			resultBuilder.build(code, result);
+
+		} catch (JSONException e) {
+			Log.e(TAG, "Invalid JSON format." + e);
```

Comment:
Shall we throw the exception upwards and then handle it appropriately within the upstream code?